File: data-structures-and-algorithm-analysis-in-c/ch02/ex-2.14.cpp
#include <cmath>
#include <iostream>

void findPrimesLessThan(const int n) {
  int *primes = new int[n + 1];

  for (int i = 2; i <= n; ++i) {
    if (primes[i] == -1) {
      continue;
    }
    std::cout << i << " ";

    if (i < sqrt(n)) {
      int j = i;

      while ((j += i) <= n) {
        primes[j] = -1;
      }
    }
  }
  std::cout << std::endl;
}

int main() {
  findPrimesLessThan(128);

  return 0;
}

File: data-structures-and-algorithm-analysis-in-c/ch02/ex-2.14.js
class LinkedList {
  constructor(data, next = null) {
    this.data = data
    this.next = next
  }

  append(data) {
    this.next = new LinkedList(data)
    return this.next
  }

  toArray() {
    const arr = []
    let node = this
    while (node && node.value !== null) {
      arr.push(node.data)
      node = node.next
    }
    return arr
  }
}

function findPrimesLessThan(n) {
  const head = new LinkedList(2)
  let node

  node = head
  let i = 1
  while (++i <= n) {
    node = node.append(i)
  }

  node = head
  while (node && node.data < Math.sqrt(n)) {
    let prev = node
    let curr = prev.next

    while (curr) {
      if (curr.data % node.data === 0) {
        prev.next = curr.next
      } else {
        prev = curr
      }

      curr = curr.next
    }

    node = node.next
  }

  return head.toArray()
}

function main() {
  console.log(findPrimesLessThan(128))
}

main()
